lisa/sut_orchestrator/libvirt/platform.py

```diff
@@ -1341,3 +1388,316 @@ def _capture_libvirt_logs(self) -> None:
+
+                assert context.pool_type
+                if "pci" in context.pool_type.value:
+                    hostdev.attrib["type"] = "pci"
+
+                    source = ET.SubElement(hostdev, "source")
+                    src_addrs = ET.SubElement(source, "address")
+
+                    assert config.domain
+                    src_addrs.attrib["domain"] = f"0x{config.domain}"
+
+                    assert config.bus
+                    src_addrs.attrib["bus"] = f"0x{config.bus}"
+
+                    assert config.slot
+                    src_addrs.attrib["slot"] = f"0x{config.slot}"
+
+                    assert config.function
+                    src_addrs.attrib["function"] = f"0x{config.function}"
+
+                    driver = ET.SubElement(hostdev, "driver")
+                    driver.attrib["name"] = "vfio"
+
+        return devices
+
+    def _get_pci_address_str(
+        self,
+        device_addr: DeviceAddressSchema,
+        with_domain: bool = False,
+    ) -> str:
+        bus = device_addr.bus
+        slot = device_addr.slot
+        fn = device_addr.function
+        domain = device_addr.domain
+        addr = f"{bus}:{slot}.{fn}"
+        if with_domain:
+            addr = f"{domain}:{addr}"
+        return addr
+
+    def _get_devices_from_pool(
+        self,
+        pool_type: HostDevicePoolType,
+        count: int,
+    ) -> List[DeviceAddressSchema]:
+        pool = self.AVAILABLE_HOST_DEVICES.get(pool_type, {})
+
+        # iommu_group: str = ""
+        # min_value = float('inf')
+        # for iommu_group_key, devices_list in pool.items():
+        #     if len(devices_list) >= count and len(devices_list) < min_value:
+        #         min_value = len(devices_list)
+        #         iommu_group = iommu_group_key
+
+        # if not iommu_group:
+        #     raise SkippedException(
+        #         f"Pool {pool_type} running out of devices: {pool}, "
+        #         "No IOMMU Group has sufficient count of devices, "
+        #         f"Refer: {pool}"
+        #     )
+
+        # # Remove the devices and update the pool
+        # # Remove entire iommu group device list so conflict does not happen
+        # devices = pool.pop(iommu_group)
+        # self.AVAILABLE_HOST_DEVICES[pool_type] = pool
+        # return devices
+
+        keys = list(pool.keys())
+        results = []
+        for r in range(1, len(keys) + 1):
+            for combo in combinations(keys, r):
+                if sum(len(pool.get(key, [])) for key in combo) == count:
+                    results.append(combo)
+        if not results:
+            for r in range(1, len(keys) + 1):
+                for combo in combinations(keys, r):
+                    if sum(len(pool.get(key, [])) for key in combo) >= count:
+                        results.append(combo)
+                        break
+                if results:
+                    break
+
+        if not results:
+            raise SkippedException(
```

Comment:
Skipped means the test condition doesn't fit, it doesn't need to test, and nothing need to be worried.

SkippedException should be used in test suite layer only. It makes test cases skipped. If it's triggered in the platform, it may cause all test cases are skipped.

The issue here is more like no resource or the environment breaks, it shouldn't be ignorable for test cases. So it's better to raise a normal exception and fail all related test cases.

Reply:
I have used `ResourceAwaitableException` as you suggested.